src/components/patterns/headingList.tsx

```diff
@@ -1,38 +1,67 @@
 "use client"
 
-import React, {useCallback, useEffect, useId, useState} from "react"
+import React, {useCallback, useEffect, useMemo, useState} from "react"
 import {twMerge} from "tailwind-merge"
 import {useDebounceCallback, useEventListener} from "usehooks-ts"
 
 type Heading = {
   text: string
   id: string
+  isDuplicateLink: boolean
+}
+
+type PageLink = {
+  text: string
+  href: string
+  context: string
 }
 
 const HeadingList = () => {
-  const uuid = useId()
   const [headings, setHeadings] = useState<Array<Heading>>([])
   const [activeHeading, setActiveHeading] = useState<string>("")
+  const [pageH1, setPageH1] = useState<string>("")
+  const [allPageLinks, setAllPageLinks] = useState<Array<PageLink>>([])
+
+  // Memoized duplicate links detection
+  const duplicateLinksMap = useMemo(() => {
+    const linkGroups = new Map<string, PageLink[]>()
+
+    // Group links by text (case-insensitive)
+    allPageLinks.forEach(link => {
+      const key = link.text.toLowerCase()
+      if (!linkGroups.has(key)) {
+        linkGroups.set(key, [])
+      }
+      linkGroups.get(key)!.push(link)
+    })
+
+    // Find duplicates (multiple hrefs or contexts)
+    const duplicates = new Set<string>()
+    linkGroups.forEach((links, text) => {
+      if (links.length > 1) {
+        const uniqueHrefs = new Set(links.map(l => l.href))
+        const uniqueContexts = new Set(links.map(l => l.context))
+        if (uniqueHrefs.size > 1 || uniqueContexts.size > 1) {
+          duplicates.add(text)
+        }
+      }
+    })
+
+    return duplicates
+  }, [allPageLinks])
```

Comment:
it should be possible to eliminate this dependency? you can scan the document on the first run, then it gets memoized for subsequent calls.

You might even be able to eliminate `allPageLinks` and the `useEffect` that builds that state.

Reply:
Ah, yes! That way makes way more sense 😮‍💨 Thanks Mike!